File: sprint4/By languages/Java/lesson10code2.java
import java.util.*;

public class Solution {
    public static Set<Triple<Integer, Integer, Integer>> effectiveSolution(int A, List<Integer> x) {
        Set<Integer> history = new HashSet<>();
        int n = x.size();
        List<Integer> xSorted = new ArrayList<>(x);
        Collections.sort(xSorted);
        Set<Triple<Integer, Integer, Integer>> triples = new HashSet<>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int target = A - xSorted.get(i) - xSorted.get(j);
                if (history.contains(target)) {
                    // Заметим, что тут тройка уже отсортирована за счёт предварительной
                    // сортировки всего массива.
                    triples.add(new Triple<>(target, xSorted.get(i), xSorted.get(j)));
                }
                history.add(xSorted.get(i));
            }
        }
        return triples;
    }
}

File: sprint4/By languages/C++/lesson10code2.cpp
#include <vector>
#include <set>
#include <unordered_set>
#include <algorithm>
#include <tuple>

std::set<std::tuple<int, int, int>> effectiveSolution(int A, const std::vector<int>& x) {
    std::unordered_set<int> history;
    int n = x.size();
    std::vector<int> xSorted(x);
    std::sort(xSorted.begin(), xSorted.end());
    std::set<std::tuple<int, int, int>> triples;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            int target = A - xSorted[i] - xSorted[j];
            if (history.count(target) > 0) {
                // Заметим, что тут тройка уже отсортирована за счёт предварительной
                // сортировки всего массива.
                triples.insert(std::make_tuple(target, xSorted[i], xSorted[j]));
            }
            history.insert(xSorted[i]);
        }
    }
    return triples;
}
